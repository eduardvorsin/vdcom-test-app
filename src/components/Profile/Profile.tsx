import React from 'react';
import { Box, SxProps, Theme, Typography, styled } from '@mui/material';

type ProfileProps = {
  name: string,
  role: string,
  src: string,
  sx?: SxProps<Theme>,
}

const ProfileWrapper = styled(Box)(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
}));

const ProfileAvatar = styled('img')(({ theme }) => ({
  flexBasis: '60px',
  height: '60px',
  objectFit: 'cover',
  borderRadius: '50%',
  border: '1px solid #7f7f7f',
  [theme.breakpoints.up('md')]: {
    marginRight: '15px',
    flexBasis: '70px',
    height: '70px',
  },
}));

const ProfileContent = styled(Box)(({ theme }) => ({
  display: 'none',
  maxWidth: '130px',
  [theme.breakpoints.up('md')]: {
    display: 'block',
  },
}));

const ProfileName = styled(Typography)(({ theme }) => ({
  fontFamily: 'inherit',
  fontSize: '18px',
  lineHeight: '1.25',
  color: '#000000',
  marginBottom: '5px',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
  [theme.breakpoints.up('md')]: {
    fontSize: '22px',
  },
}));

const ProfileRole = styled(Typography)(() => ({
  fontFamily: 'inherit',
  color: '#4d4d4d',
  fontSize: '13px',
  lineHeight: '1.2',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
}));

const Profile: React.FC<ProfileProps> = ({
  name,
  role,
  src,
  sx,
}) => (
  <ProfileWrapper
    sx={sx}
  >
    <ProfileAvatar
      src={src}
    />
    <ProfileContent>
      <ProfileName>
        {name}
      </ProfileName>
      <ProfileRole>
        {role}
      </ProfileRole>
    </ProfileContent>
  </ProfileWrapper>
);

export default Profile;

