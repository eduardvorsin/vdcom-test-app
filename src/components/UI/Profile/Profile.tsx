import React, { useEffect, useRef, useState } from 'react';
import { Avatar, Box, SxProps, Theme, Typography, styled } from '@mui/material';
import useScreenWidth from '../../../hooks/useScreenWidth/useScreenWidth';

const ProfileWrapper = styled(Box)(() => ({
  display: 'flex',
  justifyContent: 'flex-end',
  alignItems: 'center',
  position: 'relative',
}));

const ProfileAvatar = styled(Avatar)(({ theme }) => ({
  flex: '0 0 60px',
  height: '60px',
  objectFit: 'cover',
  borderRadius: '50%',
  border: '1px solid #7f7f7f',
  [theme.breakpoints.up('md')]: {
    marginRight: '15px',
    flex: '0 0 70px',
    height: '70px',
  },
}));

const ProfileName = styled(Typography)(({ theme }) => ({
  fontFamily: 'inherit',
  fontSize: '18px',
  lineHeight: '1.25',
  color: '#000000',
  marginBottom: '5px',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
  [theme.breakpoints.up('md')]: {
    fontSize: '22px',
  },
}));

const ProfileRole = styled(Typography)(() => ({
  fontFamily: 'inherit',
  color: '#4d4d4d',
  fontSize: '13px',
  lineHeight: '1.2',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
}));

const ProfileContent = styled(Box)(({ theme }) => ({
  backgroundColor: '#ddd5cf',
  position: 'absolute',
  top: 'calc(100% + 25px)',
  maxWidth: '150px',
  padding: '10px',
  transition: 'transform 0.15s ease, opacity 0.15s ease',
  opacity: '0',
  transform: 'translateY(-10px)',
  '&::before': {
    content: '""',
    position: 'absolute',
    top: '-20px',
    right: '10%',
    width: '0',
    height: '0',
    borderStyle: 'solid',
    borderWidth: '0 15px 20px 15px',
    borderColor: 'transparent transparent #ddd5cf transparent',
  },
  '&.open': {
    transition: 'transform 0.15s ease, opacity 0.15s ease',
    transform: 'translateY(0px)',
    opacity: '1',
  },
  [theme.breakpoints.up('md')]: {
    backgroundColor: 'transparent',
    padding: '0px',
    position: 'static',
    '&::before': {
      display: 'none',
    },
  },
}));

type ProfileProps = {
  name: string,
  role: string,
  src: string,
  className?: string,
  sx?: SxProps<Theme>,
  alt: string,
}

const Profile: React.FC<ProfileProps> = ({
  className,
  name,
  role,
  src,
  alt,
  sx,
}) => {
  const screenWidth = useScreenWidth();
  const isMobileWidth = screenWidth < 900;
  const [open, setOpen] = useState<boolean>(!isMobileWidth);
  const menuRef = useRef<HTMLDivElement>();

  const clickHander: React.MouseEventHandler<HTMLImageElement> = () => {
    if (isMobileWidth) {
      setOpen((prevState) => !prevState);
    }
  };

  useEffect(() => {
    const mouseDownHander = (e: MouseEvent): void => {
      if (isMobileWidth && !menuRef.current?.contains(e.target as Node)) {
        setOpen(false);
      }
    };

    document.addEventListener('mousedown', mouseDownHander);

    return () => {
      document.removeEventListener('mousedown', mouseDownHander);
    };
  }, []);

  return (
    <ProfileWrapper
      ref={menuRef}
      sx={sx}
      className={className}
      data-testid='profile'
    >
      <ProfileAvatar
        src={src}
        alt={alt}
        onClick={clickHander}
      >
        {name[0]}
      </ProfileAvatar>
      <ProfileContent
        className={open ? 'open' : ''}
      >
        <ProfileName>
          {name}
        </ProfileName>
        <ProfileRole>
          {role}
        </ProfileRole>
      </ProfileContent>
    </ProfileWrapper>
  );
};

export default Profile;
