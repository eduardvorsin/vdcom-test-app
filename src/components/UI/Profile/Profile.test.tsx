import { render, screen } from '@testing-library/react';
import React from 'react';
import Profile from './Profile';
import ProfileImg from '../../../assets/images/profile.jpg';

describe('Profile tests', () => {
  it('should render correctly', () => {
    render(
      <Profile
        alt='Ralph James avatar'
        name='Ralph James'
        role='CEO'
        src={ProfileImg}
      />,
    );

    expect(screen.getByTestId<HTMLDivElement>('profile')).toBeInTheDocument();
  });

  it('should correctly apply className to the element', () => {
    render(
      <Profile
        alt='Ralph James'
        name='Ralph James avatar'
        role='CEO'
        src={ProfileImg}
        className='profile'
      />,
    );

    expect(screen.getByTestId<HTMLDivElement>('profile')).toHaveClass('profile');
  });

  it('should correctly display the passed values for the role and name propses', () => {
    render(
      <Profile
        alt='Sam Evans avatar'
        role='Manager'
        name='Sam Evans'
        src={ProfileImg}
      />,
    );

    expect(screen.getByText<HTMLParagraphElement>('Sam Evans')).toBeInTheDocument();
    expect(screen.getByText<HTMLParagraphElement>('Manager')).toBeInTheDocument();
  });

  it('should correctly display the image with the passed src address', () => {
    render(
      <Profile
        alt='Sam Evans avatar'
        role='Manager'
        name='Sam Evans'
        src={ProfileImg}
      />,
    );

    expect(screen.getByAltText<HTMLImageElement>('Sam Evans avatar')).toBeInTheDocument();
  });

  it('is a basic snapshot', () => {
    render(
      <Profile
        alt='Adam Richards avatar'
        role='Designer'
        name='Adam Richards'
        src={ProfileImg}
      />,
    );

    expect(screen.getByTestId<HTMLDivElement>('profile')).toMatchSnapshot();
  });
});
