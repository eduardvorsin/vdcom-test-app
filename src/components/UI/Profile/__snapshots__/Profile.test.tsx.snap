// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile tests is a basic snapshot 1`] = `
<div
  class="MuiBox-root css-ro3nde"
  data-testid="profile"
>
  <div
    class="MuiAvatar-root MuiAvatar-circular css-quq592-MuiAvatar-root"
  >
    <img
      alt="Adam Richards avatar"
      class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
      src="assets"
    />
  </div>
  <div
    class=" MuiBox-root css-yi99sy"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-14tr2nh-MuiTypography-root"
    >
      Adam Richards
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-j1f0n7-MuiTypography-root"
    >
      Designer
    </p>
  </div>
</div>
`;
